### PR TITLE
Detect newer Visual Studio compilers correctly like VS2017.

### PR DESCRIPTION
**Pull request by brenttaylor**
This is just a cherry-picked commit.  Adds proper detection of VS2017 so we're not locked to VS2015.

### CODE REVIEW

File: methods.py [hunk 1]
```diff
@@ -1549,6 +1549,33 @@ def detect_visual_c_compiler_version(tools_env):
             vc_chosen_compiler_index = vc_x86_amd64_compiler_detection_index
             vc_chosen_compiler_str = "x86_amd64"
 
+    # Newer versions have a different path available
+    vc_amd64_compiler_detection_index = tools_env["PATH"].upper().find(tools_env['VCTOOLSINSTALLDIR'].upper() + "BIN\\HOSTX64\\X64;")
```
Comment (vnen) [inline methods.py]: This fails if `VCTOOLSINSTALLDIR` is not set, which is the case for versions before 2017. Add the changes from 8b01f9d28ffccb21db04867d1265f0be9d40aa53 to make it work properly.

**Review by vnen — CHANGES_REQUESTED** [review]
This will break older toolsets.

File: methods.py (hunk 1 above)
Comment (brenttaylor) [inline methods.py]: Oof, good catch.  I'll get that done today.

Comment (akien-mga) [inline methods.py]: Bump.

**Comment by akien-mga** [comment]
Closing as it is getting stale. Feel free to open a new PR with the necessary changes mentioned by @vnen above.

**Pull request closed without merging**